[WebGPU] popErrorScope causes an IPC assertion

mwyrzykowski opened the pull request:
#### d5a1264b45a7b6e3e77c6d0017090a965e63ae91
<pre>
[WebGPU] popErrorScope causes an IPC assertion
<a href="https://bugs.webkit.org/show_bug.cgi?id=251666">https://bugs.webkit.org/show_bug.cgi?id=251666</a>
&lt;radar://104993240&gt;

Reviewed by Myles C. Maxfield.

GPUDevice.popErrorScope is supposed to be asynchronous, so use sendWithAsyncReply
instead of sendSync.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::popErrorScope):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:

Canonical link: <a href="https://commits.webkit.org/260413@main">https://commits.webkit.org/260413@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88020951684a3aee14e4e27c5a996243c95f74b5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41054 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117311 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116626 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8556 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100395 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113955 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41972 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95986 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28902 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10133 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30249 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10858 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7154 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16282 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49841 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7202 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12440 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -287,20 +287,16 @@ static void captureFrame(id<MTLDevice> captureObject)
     // https://gpuweb.github.io/gpuweb/#dom-gpudevice-poperrorscope
 
     if (!validatePopErrorScope()) {
-        instance().scheduleWork([callback = WTFMove(callback)]() mutable {
```
Ditto for the other patch. This function philosophically should be async; we should be making it properly async rather than temporarily making it sync.

OK for temporary code, but I think we can do better.

mwyrzykowski replied:
Why should it be async and doesn't scheduleWork() run on the same dispatch queue which the containing function runs? If so, what is the benefit to making it async?

litherum replied:
Because that's the API contract of promises, you are guaranteed your promise's `.then` callback won't run in the middle of your `popErrorScope()` call. I'm trying to match the API behavior of the WebGPU IDL as much as possible.

mwyrzykowski replied:
That's correct but we shouldn't impact the web api for due to how the native API is structured. Instead I think we should ensure the contract on the native API in a way that does not impact the web API.

litherum replied:
I'm confused. An app that wants to use WebGPU.framework is going to expect that it behaves the same way as the web API does, right?

mwyrzykowski replied:
I think so however native WebGPU still seems under development. 

My point is that making the callback asynchronous here negatively impacts users of the web api as now they incur additional latency for the task to return, when it was already asynchronous from the website’s perspective. 

If we want the native API to appear the same, we should implement that in a way that doesn’t add latency to the web api. This patch doesn’t address that as the web api is priority.

litherum replied:
Is the difference measurable?

litherum replied:
If the answer is "yes", I propose we add a boolean to `WGPUInstanceCocoaDescriptor` that represents whether violating the asynchronicity guarantees of the spec is more desirable than upholding them, and set the value of the boolean to true in WebKit and false everywhere else, and then implement both codepaths, and use the boolean to select between them.

If the answer is "no" then we should just use the async codepath everywhere.

mwyrzykowski replied:
Probably not measurable especially given the discussion is about `popErrorScope`

I will try the new IPC `sendToStreamWithAsyncReply` to see if we can resolve this issue that way. I'm kind of opposed to using NotStreamEncodableReply as that is a workaround

mwyrzykowski replied:
@litherum PR updated with `sendWithAsyncReply`, I reverted the changes to `Device.mm`

mwyrzykowski committed:
[WebGPU] popErrorScope causes an IPC assertion
https://bugs.webkit.org/show_bug.cgi?id=251666
<radar://104993240>

Reviewed by Myles C. Maxfield.

GPUDevice.popErrorScope is supposed to be asynchronous, so use sendWithAsyncReply
instead of sendSync.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::popErrorScope):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:

Canonical link: https://commits.webkit.org/260413@main
